compiler+runtime/src/cpp/jank/runtime/obj/ratio.cpp

```diff
@@ -0,0 +1,556 @@
+  {
+    int gcd = std::gcd(numerator, denominator);
+    numerator /= gcd;
+    denominator /= gcd;
+
+    if(denominator < 0)
+    {
+      numerator = -1 * numerator;
+      denominator = -1 * denominator;
+    }
+  }
+
+  void obj::ratio::to_string(fmt::memory_buffer &buff) const
+  {
+    format_to(std::back_inserter(buff), FMT_COMPILE("{}/{}"), data.numerator, data.denominator);
+  }
+
+  native_persistent_string obj::ratio::to_string() const
+  {
+    fmt::memory_buffer buff;
+    to_string(buff);
+    return native_persistent_string{ buff.data(), buff.size() };
+  }
+
+  native_persistent_string obj::ratio::to_code_string() const
+  {
+    return to_string();
+  }
+
+  native_hash obj::ratio::to_hash() const
+  {
+    return hash::combine(hash::integer(data.numerator), hash::integer(data.denominator));
+  }
+
+  native_bool obj::ratio::equal(object const &o) const
+  {
+    if(o.type == object_type::integer)
+    {
+      auto const i(expect_object<obj::integer>(&o));
+      return this->data == i->data;
+    }
+
+    if(o.type == object_type::real)
+    {
+      auto const i(expect_object<obj::integer>(&o));
+      return this->data == i->data;
+    }
+
+    if(o.type == object_type::ratio)
+    {
+      return this == expect_object<obj::ratio>(&o).data;
+    }
+
+    return false;
+  }
+
+  native_integer obj::ratio::compare(object const &o) const
+  {
+    return visit_number_like(
+      [this](auto const typed_o) -> native_integer {
+        return (data > typed_o->data) - (data < typed_o->data);
+      },
+      [&]() -> native_integer {
+        throw std::runtime_error{ fmt::format("not comparable: {}", runtime::to_string(&o)) };
+      },
+      &o);
+  }
+
+  native_integer obj::ratio::compare(static_object const &o) const
+  {
+    return (data > o.data) - (data < o.data);
+  }
+
+  object_ptr operator+(obj::ratio_data const l, obj::ratio_data const r)
+  {
+    native_integer denom = l.denominator * r.denominator;
+    native_integer num = l.numerator * r.denominator + r.numerator * l.denominator;
+    return obj::ratio::create(num, denom);
+  }
+
+  obj::ratio_ptr operator+(obj::integer_ptr l, obj::ratio_data const r)
+  {
+    return expect_object<obj::ratio>(obj::ratio_data(l->data, 1LL) + r);
+  }
+
+  obj::ratio_ptr operator+(obj::ratio_data const l, obj::integer_ptr r)
+  {
+    return r + l;
+  }
+
+  native_real operator+(obj::real_ptr l, obj::ratio_data const r)
+  {
+    return l->data + r.to_real();
+  }
+
+  native_real operator+(obj::ratio_data const l, obj::real_ptr r)
+  {
+    return l.to_real() + r->data;
+  }
+
+  native_real operator+(obj::ratio_data const l, native_real r)
+  {
+    return l.to_real() + r;
+  }
+
+  native_real operator+(native_real l, obj::ratio_data const r)
+  {
+    return l + r.to_real();
+  }
+
+  obj::ratio_ptr operator+(obj::ratio_data const l, native_integer r)
+  {
+    return expect_object<obj::ratio>(
+      obj::ratio::create(l.numerator + r * l.denominator, l.denominator));
```

Comment:
Are we 100% certain that calling `ratio::create` here will give us a ratio? If so, we should just make a ratio.

Reply:
yes, a `ratio_data` exists only if the denominator is not 1 (its ctor made sure of that). so a ratio + int will always be another ratio. 

updated to be more efficient.